core/optic/shared/src/main/scala/com/useoptic/diff/initial/DistributionAwareShapeBuilder.scala

```diff
@@ -0,0 +1,265 @@
+            case ShapesHelper.NumberKind => PrimitiveKind(ShapesHelper.NumberKind, trail, idGenerator())
+            case ShapesHelper.BooleanKind => PrimitiveKind(ShapesHelper.BooleanKind, trail, idGenerator())
+            case ShapesHelper.NullableKind => {
+              val notNullExamples = examples.filterNot(_.isNull)
+              val innerNull = fromJsons(notNullExamples, trail, true, examples.size)
+              NullableShape(innerNull, trail, idGenerator())
+            }
+          }
+        }
+      }.toSeq
+
+      def flattenShapes(shapes: Seq[ShapesToMake]): ShapesToMake = {
+        if (shapes.isEmpty) {
+          Unknown(trail, idGenerator())
+        } else if (shapes.size == 1) {
+          shapes.head
+        } else {
+          if (shapes.exists(_.isInstanceOf[NullableShape])) {
+            //override with nullable
+            val remainingShapes = shapes.filterNot(_.isInstanceOf[NullableShape])
+            NullableShape(flattenShapes(remainingShapes), trail, idGenerator())
+          } else {
+            OneOfShape(shapes, trail, idGenerator())
+          }
+        }
+      }
+
+      flattenShapes(shapesToMake)
+    }
+  }
+
+
+}
+
+
+
+//// Shapes to Make
+trait ShapesToMake {
+  def id: String
+  def trail: JsonTrail
+}
+
+case class OptionalShape(shape: ShapesToMake, trail: JsonTrail, id: String) extends ShapesToMake
+case class NullableShape(shape: ShapesToMake, trail: JsonTrail, id: String) extends ShapesToMake
+case class OneOfShape(branches: Seq[ShapesToMake], trail: JsonTrail, id: String) extends ShapesToMake
+case class ObjectWithFields(fields: Seq[FieldWithShape], trail: JsonTrail, id: String) extends ShapesToMake
+case class ListOfShape(shape: ShapesToMake, trail: JsonTrail, id: String) extends ShapesToMake
+case class FieldWithShape(key: String, shape: ShapesToMake, trail: JsonTrail, id: String) extends ShapesToMake
+case class PrimitiveKind(baseShape: CoreShapeKind, trail: JsonTrail, id: String) extends ShapesToMake
+case class Unknown(trail: JsonTrail, id: String) extends ShapesToMake
+////
+
+class TrailValueMap(val totalSamples: Int) {
+  private val _internal = scala.collection.mutable.Map[JsonTrail, Vector[Option[JsonLike]]]()
+
+  def putValue(trail: JsonTrail, value: JsonLike): Unit = putValueOptional(trail, Some(value))
+
+  def putValueOptional(trail: JsonTrail, value: Option[JsonLike]): Unit = {
+    if (_internal.contains(trail)) {
+      _internal.put(trail, _internal(trail) :+ value)
+    } else {
+      _internal.put(trail, Vector(value))
+    }
+  }
+
+  def getRoot = valuesForTrail(JsonTrail(Seq.empty))
+
+  def toMap: Map[JsonTrail, Vector[Option[JsonLike]]] = _internal.toMap
+
+  def valuesForTrail(trail: JsonTrail): Vector[Option[JsonLike]] = _internal.getOrElse(trail, Vector.empty)
+
+  def hasTrail(trail: JsonTrail) = _internal.contains(trail)
+
+}
+
+class ShapeBuilderVisitor(aggregator: TrailValueMap) extends JsonLikeVisitors {
```

Comment:
We should use the JsonLikeAndSpecVisitors from now on - let me know if you want a hand migrating

Reply:
Let’s do it during our session tomorrow. Didn’t realize it was new!